pkg/pillar/zedbox/zedbox.go

```diff
@@ -32,45 +37,155 @@ import (
-	"command":          command.Run,
-	"diag":             diag.Run,
-	"domainmgr":        domainmgr.Run,
-	"downloader":       downloader.Run,
-	"executor":         executor.Run,
-	"hardwaremodel":    hardwaremodel.Run,
-	"ledmanager":       ledmanager.Run,
-	"logmanager":       logmanager.Run,
-	"nim":              nim.Run,
-	"nodeagent":        nodeagent.Run,
-	"verifier":         verifier.Run,
-	"volumemgr":        volumemgr.Run,
-	"waitforaddr":      waitforaddr.Run,
-	"zedagent":         zedagent.Run,
-	"zedmanager":       zedmanager.Run,
-	"zedrouter":        zedrouter.Run,
-	"ipcmonitor":       ipcmonitor.Run,
-	"baseosmgr":        baseosmgr.Run,
-	"wstunnelclient":   wstunnelclient.Run,
-	"conntrack":        conntrack.Run,
-	"tpmmgr":           tpmmgr.Run,
-	"vaultmgr":         vaultmgr.Run,
-	"upgradeconverter": upgradeconverter.Run,
-}
+const (
+	agentName   = "zedbox"
+	socketPath  = "/var/run"
+	errorTime   = 3 * time.Minute
+	warningTime = 40 * time.Second
+)
+
+var (
+	debugOverride = false
+	//Version Set from Makefile
+	Version     = "No version specified"
+	entrypoints = map[string]func(*pubsub.PubSub){
+		"client":           client.Run,
+		"command":          command.Run,
+		"diag":             diag.Run,
+		"domainmgr":        domainmgr.Run,
+		"downloader":       downloader.Run,
+		"executor":         executor.Run,
+		"hardwaremodel":    hardwaremodel.Run,
+		"ledmanager":       ledmanager.Run,
+		"logmanager":       logmanager.Run,
+		"nim":              nim.Run,
+		"nodeagent":        nodeagent.Run,
+		"verifier":         verifier.Run,
+		"volumemgr":        volumemgr.Run,
+		"waitforaddr":      waitforaddr.Run,
+		"zedagent":         zedagent.Run,
+		"zedmanager":       zedmanager.Run,
+		"zedrouter":        zedrouter.Run,
+		"ipcmonitor":       ipcmonitor.Run,
+		"baseosmgr":        baseosmgr.Run,
+		"wstunnelclient":   wstunnelclient.Run,
+		"conntrack":        conntrack.Run,
+		"tpmmgr":           tpmmgr.Run,
+		"vaultmgr":         vaultmgr.Run,
+		"upgradeconverter": upgradeconverter.Run,
+	}
+)
 
 func main() {
+	log.Infof("Starting %s", agentName)
+	versionPtr := flag.Bool("v", false, "Version")
+	debugPtr := flag.Bool("d", false, "Debug flag")
+
 	ps := pubsub.New(&socketdriver.SocketDriver{})
+	var sktData string
+	var cmdAndArgs []string
+	sktChan := make(chan string)
+	stillRunning := time.NewTicker(15 * time.Second)
+	dumpHeapMemory := time.NewTicker(5 * time.Second)
+	debugOverride = *debugPtr
 
-	basename := filepath.Base(os.Args[0])
-	if entrypoint, ok := entrypoints[basename]; ok {
-		entrypoint(ps)
+	if *versionPtr {
+		fmt.Printf("%s: %s\n", os.Args[0], Version)
+		return
+	}
+	if debugOverride {
+		log.SetLevel(log.DebugLevel)
 	} else {
-		fmt.Printf("Unknown package: %s\n", basename)
-		os.Exit(1)
+		log.SetLevel(log.InfoLevel)
+	}
+
+	agentlog.Init(agentName)
+	if err := pidfile.CheckAndCreatePidfile(agentName); err != nil {
+		log.Fatal(err)
+	}
+
+	go connectAndRead(agentName, sktChan)
```

Comment:
Instead of partial duplication of pubsub, would it make sense to define a service+args golang struct and use that with pubsub to tell zedbox what to spawn? That assumes that the code which starts the services is a golang app (perhaps a tiny one only importing pubsub).